File: ios/Classes/MoproFlutterPlugin.swift
import Flutter
import UIKit
import moproFFI


public class MoproFlutterPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "mopro_flutter", binaryMessenger: registrar.messenger(), codec: FlutterJSONMethodCodec())
    let instance = MoproFlutterPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "generateProof":
      guard let args = call.arguments as? [String: Any],
            let zkeyPath = args["zkeyPath"] as? String,
            let inputs = args["inputs"] as? [String: [String]] else {
        result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing arguments", details: nil))
        return
      }

      do {
        // Call the function from mopro.swift
        let proofResult = try generateCircomProof(zkeyPath: zkeyPath, circuitInputs: inputs)
        
        // Convert proof and inputs to Base64 strings
        let proofBase64 = proofResult.proof.base64EncodedString()
        let inputsBase64 = proofResult.inputs.base64EncodedString()

        // Build the response
        let response: [String: String] = [
          "proof": proofBase64,
          "inputs": inputsBase64
        ]

        // Return the response to Flutter
        result(response)
      } catch {
        result(FlutterError(code: "PROOF_GENERATION_ERROR", message: "Failed to generate proof", details: error.localizedDescription))
      }

    default:
      result(FlutterMethodNotImplemented)
    }
  }
}

File: android/src/main/kotlin/com/example/mopro_flutter/MoproFlutterPlugin.kt
package com.example.mopro_flutter

import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

import android.content.Context
import kotlinx.coroutines.launch
import uniffi.mopro.generateCircomProof
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

import org.json.JSONObject
import android.util.Base64
import android.util.Log
import io.flutter.plugin.common.JSONMethodCodec
import uniffi.mopro.GenerateProofResult

/** MoproFlutterPlugin */
class MoproFlutterPlugin: FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "mopro_flutter", JSONMethodCodec.INSTANCE)
    channel.setMethodCallHandler(this)
  }

  fun JSONObject.toMap(): Map<String, List<String>> {
    val map = mutableMapOf<String, List<String>>()
    this.keys().forEach { key ->
      val list = mutableListOf<String>()
      val jsonArray = this.getJSONArray(key)
      for (i in 0 until jsonArray.length()) {
        list.add(jsonArray.getString(i))
      }
      map[key] = list
    }
    return map
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    if (call.method == "generateProof") {
      val zkeyPath = call.argument<String>("zkeyPath") ?: return result.error("ARGUMENT_ERROR", "Missing zkeyPath", null)

      val inputsJson = call.argument<JSONObject>("inputs") ?: return result.error("ARGUMENT_ERROR", "Missing inputs", null)
      val inputs = inputsJson.toMap() as Map<String, List<String>>
 
      val res: GenerateProofResult = generateCircomProof(zkeyPath, inputs)

      Log.d("generateProof", "$res")
      // Convert ByteArray fields to Base64 strings
        val proofBase64 = Base64.encodeToString(res.proof, Base64.NO_WRAP)
        val inputsBase64 = Base64.encodeToString(res.inputs, Base64.NO_WRAP)

        // Build the JSON response
        val json = JSONObject().apply {
            put("proof", proofBase64)  // Base64-encoded proof
            put("inputs", inputsBase64)  // Base64-encoded inputs
        }

        // Send the JSON string back to Flutter
        result.success(
          json
        )
    } else {
      result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}
